### PR TITLE
Timestamp added to tcpreassembly and OOO buffer clean condition modified

### PR DESCRIPTION
**Pull request by HrshWardhan**
Added Timestamp field to TcpStreamData for this information to be available in callbacks as it might be necessary for certain use cases. 

Adding a config parameter in TcpReassemblyConfiguration which disables the Out of order buffer clean condition when a data packet from opposite side arrives, Since there are instances where this buffer clear condition can lead to declaration of excessive missing packets. Hence user should have a config file parameter  to disable this and purely rely on max buffer size condition. As none of them are perfect solutions this will give a little more control over it.

### CODE REVIEW

**Comment by seladb** [comment]
@HrshWardhan the build is failing and has conflicts. Do you want to fix them or should we close this PR?

**Comment by HrshWardhan** [comment]
I am working on fixing them, got sidetracked a little, will be done soon.

**Comment by seladb** [comment]
@HrshWardhan this PR has been open for quite some time. Should we close it?

**Comment by HrshWardhan** [comment]
@seladb I have resolved the merge and build conflicts. Sorry for the huge delay.

File: Packet++/src/TcpReassembly.cpp [hunk 1]
```diff
@@ -159,6 +160,7 @@ TcpReassembly::ReassemblyStatus TcpReassembly::reassemblePacket(Packet& tcpData)
 		}
 	}
 
+	timeval timestampOfTheRecievedPacket = timespecToTimeval(tcpData.getRawPacket()->getPacketTimeStamp());
```
Comment (seladb) [inline Packet++/src/TcpReassembly.cpp:163]: we already fetch the packet timestamp if the connection isn't new [here](https://github.com/seladb/PcapPlusPlus/blob/d8d702acc99bf8bbf1531f00a3162c6b11a3cd2c/Packet%2B%2B/src/TcpReassembly.cpp#L147). Maybe we should just move it to after [this line](https://github.com/seladb/PcapPlusPlus/blob/d8d702acc99bf8bbf1531f00a3162c6b11a3cd2c/Packet%2B%2B/src/TcpReassembly.cpp#L113) so we don't call the same method twice?

File: Packet++/header/TcpReassembly.h [hunk 2]
```diff
@@ -197,16 +205,26 @@ struct TcpReassemblyConfiguration
...
 	 */
-	TcpReassemblyConfiguration(bool removeConnInfo = true, uint32_t closedConnectionDelay = 5, uint32_t maxNumToClean = 30, uint32_t maxOutOfOrderFragments = 0) :
-		removeConnInfo(removeConnInfo), closedConnectionDelay(closedConnectionDelay), maxNumToClean(maxNumToClean), maxOutOfOrderFragments(maxOutOfOrderFragments)
+	TcpReassemblyConfiguration(bool removeConnInfo = true, uint32_t closedConnectionDelay = 5, uint32_t maxNumToClean = 30, uint32_t maxOutOfOrderFragments = 0, 
+		bool enableBaseBufferClearCondtion = true) : removeConnInfo(removeConnInfo), closedConnectionDelay(closedConnectionDelay), maxNumToClean(maxNumToClean), maxOutOfOrderFragments(maxOutOfOrderFragments), enableBaseBufferClearCondtion(enableBaseBufferClearCondtion)
 	{
+		if(!enableBaseBufferClearCondtion && maxOutOfOrderFragments==0)
+		{
+			// set a default value to prevent indefinate filling of out of order buffer 
+			maxOutOfOrderFragments = 20;
```
Comment (seladb) [inline Packet++/header/TcpReassembly.h:226]: I'm not sure we need a default value here. If users set the value of` `maxOutOfOrderFragments` to be `0` they expect the buffer to never be cleaned. Setting a value for them is kind of weird behavior. WDYT?

File: Tests/Pcap++Test/TestDefinition.h [hunk 3]
```diff
@@ -65,6 +65,8 @@ PTF_TEST_CASE(TestTcpReassemblyIPv6_OOO);
 PTF_TEST_CASE(TestTcpReassemblyCleanup);
 PTF_TEST_CASE(TestTcpReassemblyMaxOOOFrags);
 PTF_TEST_CASE(TestTcpReassemblyMaxSeq);
+PTF_TEST_CASE(TestTcpReassemblyDBOOOBufferCC);
```
Comment (seladb) [inline Tests/Pcap++Test/TestDefinition.h:68]: Too many initials 😄 
what do you think about: `TestTcpReassemblyDisableOOOCleanup`

File: Tests/Pcap++Test/TestDefinition.h [hunk 4]
```diff
@@ -65,6 +65,8 @@ PTF_TEST_CASE(TestTcpReassemblyIPv6_OOO);
...
+PTF_TEST_CASE(TestTcpReassemblyTstamps);
```
Comment (seladb) [inline Tests/Pcap++Test/TestDefinition.h:69]: Maybe change to `TestTcpReassemblyTimestamps`?

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 5]
```diff
@@ -1054,3 +1056,68 @@ PTF_TEST_CASE(TestTcpReassemblyMaxSeq)
...
+	PTF_ASSERT_EQUAL(stats1.size(), 1);
+	PTF_ASSERT_EQUAL(stats2.size(), 1);
+	PTF_ASSERT_EQUAL(stats1.begin()->second.numOfDataPackets, 18);
+	PTF_ASSERT_EQUAL(stats2.begin()->second.numOfDataPackets, 19);
+
+	packetStream.clear();
+	tcpReassemblyResults.clear();
+} // TestTcpReassemblyDBOOOBufferCC
+
+PTF_TEST_CASE(TestTcpReassemblyTstamps) // TestTcpReassemblyTimeStamps
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1098]: If we change the test name I don't think we need the `// TestTcpReassemblyTimeStamps` comment

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 6]
```diff
@@ -1054,3 +1056,68 @@ PTF_TEST_CASE(TestTcpReassemblyMaxSeq)
...
+
+	TcpReassemblyMultipleConnStats::Stats &stats = tcpReassemblyResults.stats;
+	PTF_ASSERT_EQUAL(stats.begin()->second.numOfDataPackets,7);
+	std::ifstream expectedOutput("PcapExamples/timestamp_output.txt");
+	for(long unsigned int i = 0;i<tcpReassemblyResults.timestamps.size();i++){
+		timeval t = tcpReassemblyResults.timestamps[i];
+		std::string expected;
+		expectedOutput>>expected;
+		int expUsec = std::stoll(expected)%1000000;
+		int expSec = std::stoll(expected)/1000000;
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1116]: `std:stoll` is C++11 syntax which PcapPlusPlus hasn't yet officially moved to. We plan to work on it in the coming months. Can we use `atoll` for now and put `std:stoll` in a `TODO C++11` comment?

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 7]
```diff
@@ -1054,3 +1056,68 @@ PTF_TEST_CASE(TestTcpReassemblyMaxSeq)
...
 } //TestTcpReassemblyMaxSeq
+
+PTF_TEST_CASE(TestTcpReassemblyDBOOOBufferCC) // TestTcpReassemblyDisableBaseOutOfOrderBufferClearCondition
+{
+	std::string errMsg;
+	std::vector<pcpp::RawPacket> packetStream;
+	TcpReassemblyMultipleConnStats results1;
+	TcpReassemblyMultipleConnStats results2;
+	pcpp::TcpReassemblyConfiguration config1(true, 5, 30,20,true);
+	pcpp::TcpReassemblyConfiguration config2(true, 5, 30,20,false);
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1067]: missing a few spaces:
```cpp
pcpp::TcpReassemblyConfiguration config1(true, 5, 30, 20, true);
pcpp::TcpReassemblyConfiguration config2(true, 5, 30, 20, false);
```

File: Packet++/src/TcpReassembly.cpp (hunk 1 above)
Comment (HrshWardhan) [reply Packet++/src/TcpReassembly.cpp:163]: Should i replace the call in [this](https://github.com/seladb/PcapPlusPlus/blob/d8d702acc99bf8bbf1531f00a3162c6b11a3cd2c/Packet%2B%2B/src/TcpReassembly.cpp#L128) line too?

Comment (seladb) [reply Packet++/src/TcpReassembly.cpp:163]: Yes, I think so. Thanks for pointing this out, I missed it... :)

File: Packet++/header/TcpReassembly.h (hunk 2 above)
Comment (HrshWardhan) [reply Packet++/header/TcpReassembly.h]: Yeah setting a default value will be bad according to previous functionality of maxOutOfOrderFrafments. I have removed this

**Comment by HrshWardhan** [comment]
@seladb I have integrated all of the above mentioned suggestions. Is there some issue with continuous-integration/appveyor/pr ?  As it is passing when i re-ran it on my fork. Can you look into this?

**Comment by seladb** [comment]
> @seladb I have integrated all of the above mentioned suggestions. Is there some issue with continuous-integration/appveyor/pr ? As it is passing when i re-ran it on my fork. Can you look into this?

Yes, we recently started running live network tests in the CI so it is still not 100% stable. We'll address these issues soon. I re-ran the failing test and it passed. I think we're ready to merge this PR!

BIG THANK YOU for your contribution to this project, we really appreciate it! 🙏 🙏

**Pull request merged**